File: server/models/AuditLog.js
const mongoose = require("mongoose");

const auditLogSchema = new mongoose.Schema(
  {
    action: {
      type: String,
      required: true,
      enum: [
        "user_registered", "user_login", "user_logout",
        "msme_created", "msme_updated", "msme_deleted",
        "gst_uploaded", "transaction_uploaded",
        "score_generated",
        "loan_submitted", "loan_approved", "loan_rejected",
        "document_uploaded", "document_verified",
      ],
    },
    performedBy: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
    targetMsmeId: { type: mongoose.Schema.Types.ObjectId, ref: "MSME" },
    entityType: { type: String },
    entityId: { type: mongoose.Schema.Types.ObjectId },
    ipAddress: { type: String },
    userAgent: { type: String },
    payload: { type: mongoose.Schema.Types.Mixed },
  },
  { timestamps: true }
);

auditLogSchema.index({ targetMsmeId: 1 });
auditLogSchema.index({ action: 1 });
auditLogSchema.index({ performedBy: 1 });
auditLogSchema.index({ createdAt: -1 });

module.exports = mongoose.model("AuditLog", auditLogSchema);

File: server/app/models/audit_log.py
"""AuditLog document — mirrors the original Mongoose AuditLog schema."""

from beanie import Document
from pydantic import Field
from typing import Optional, Literal, Any
from datetime import datetime


class AuditLog(Document):
    action: Literal[
        "user_registered", "user_login", "user_logout",
        "msme_created", "msme_updated", "msme_deleted",
        "gst_uploaded", "transaction_uploaded",
        "score_generated",
        "loan_submitted", "loan_approved", "loan_rejected",
        "document_uploaded", "document_verified",
    ]
    performed_by: Optional[str] = None  # User ID
    target_msme_id: Optional[str] = None  # MSME ID
    entity_type: Optional[str] = None
    entity_id: Optional[str] = None
    ip_address: Optional[str] = "unknown"
    user_agent: Optional[str] = "unknown"
    payload: Optional[Any] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "auditlogs"
